modules/puppet/templates/puppet-darwin-run-puppet.sh.erb

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+PUPPET_ENV='production'
+PUPPET_REPO="${PUPPET_REPO:-<%= @puppet_repo -%>}"
+PUPPET_BRANCH="${PUPPET_BRANCH:-<%= @puppet_branch -%>}"
+WORKING_DIR="/etc/puppet/environments/${PUPPET_ENV}/code"
+R10K_DIR="/etc/puppet/environments/${PUPPET_ENV}/r10k_modules"
+ROLE_FILE='/etc/puppet_role'
+PUPPET_BIN='/opt/puppetlabs/bin/puppet'
+FACTER_BIN='/opt/puppetlabs/bin/facter'
+R10K_BIN='/opt/puppetlabs/bin/r10k'
+FQDN=$(${FACTER_BIN} networking.fqdn)
+
+export LANG=en_US.UTF-8
+
+function fail {
+    # TODO: report failure to ext service
+    echo "${@}"
+    exit 1
+}
+
+mkdir -p "${WORKING_DIR}" || fail
+mkdir -p "${R10K_DIR}" || fail
+cd "${WORKING_DIR}" || fail
+
+# Make dir world writable for puppet homebrew bug workaround
+chmod 777 .
+
+# Set role or fail if file not found
+if [ -f "${ROLE_FILE}" ]; then
+    ROLE=$(<${ROLE_FILE})
+else
+    fail "Failed to find puppet role file ${ROLE_FILE}"
+fi
+
+[ -f '/var/root/vault.yaml' ] || fail "Secrets file not found"
+
+function update_puppet {
+
+    # Initialize working dir if dir is empty
+    if [ ! "$(find "$WORKING_DIR" -mindepth 1 -print -quit 2>/dev/null)" ]; then
+        git init || return 1
+        git remote add origin "${PUPPET_REPO}" || return 1
```

Comment:
Because cloning does an initial pull of the origin/master into the working directory and isn't needed.  What is needed follows with a fetch of all branches and then a checkout which results in a detached head.  That is what we want.

Reply:
Okay I didn't know it needed to be a detached head. If it doesn't need to be detached, could it be simpler to always do a sparse/shallow clone of the explicit branch? That would create the directory, fetch the latest, and only get the explicit branch, and not pull all the history.